src/plots/cartesian/set_convert.js

```diff
@@ -182,8 +182,9 @@ module.exports = function setConvert(ax) {
             // encounters them, ie all the categories from the
             // first data set, then all the ones from the second
             // that aren't in the first etc.
-            // TODO: sorting options - do the sorting
-            // progressively here as we insert?
+            // it is assumed that this function is being invoked in the
+            // already sorted category order; otherwise there would be
+            // a disconnect between the array and the index returned
```

Comment:
@monfera Thanks for this very detailed overview.

You've got me convinced, `ax._categories` can't be sorted from inside the `d2c` routine.

We need to fill in `ax._categories` earlier. I think within [`makeCalcdata`](https://github.com/plotly/plotly.js/blob/e03e9e0bd7ea3a6bed3b32bd784276e8e9eb70f1/src/plots/cartesian/set_convert.js#L208-L230) makes the most sense at the moment.